airflow/hooks/hdfs_hook.py

```diff
@@ -17,85 +17,121 @@
 
+    By default hdfs3 loads its configuration from `core-site.xml` and
+    `hdfs-site.xml` if these files can be found in any of the typical
+    locations. The hook loads `host` and `port` parameters from the
+    hdfs connection (if given) and extra configuration parameters can be
+    supplied in the connections extra JSON as follows:
 
-class HDFSHookException(AirflowException):
-    pass
+    {
+        "pars": {
+            "dfs.domain.socket.path": "/var/lib/hadoop-hdfs/dn_socket"
+        },
+        "ha": {
+            "host": "nameservice1",
+            "conf": {
+                "dfs.nameservices": "ns1",
+                "dfs.ha.namenodes.ns1": "nn1,nn2",
+                "dfs.namenode.rpc-address.ns1.nn1": "host1:8020",
+                "dfs.namenode.rpc-address.ns1.nn2": "host2:8020",
+                "dfs.namenode.http-address.ns1.nn1": "host1:50070",
+                "dfs.namenode.http-address.ns1.nn2": "host2:50070"
+            }
+        },
+        "autoconf": True,
+        "token": "...",
+        "ticket_cache": "..."
+    }
 
+    Here `pars` can be used to supply configuration options with the same key
+    names as typically contained in the XML config files, which will take
+    precedence over any parameters loaded from files. The `ha` configuration
+    section can be used to supply options for using hdfs3 in high-availability
+    mode (see the hdfs3 documentation for more details). The `autoconf` option
+    controls whether hdfs3 uses the available hadoop config files for
+    its configuration, whilst the `token` and `ticket_cache` options are
+    used for configuring kerberos.
 
-class HDFSHook(BaseHook):
-    """
-    Interact with HDFS. This class is a wrapper around the snakebite library.
-
-    :param hdfs_conn_id: Connection id to fetch connection info
-    :type conn_id: string
-    :param proxy_user: effective user for HDFS operations
-    :type proxy_user: string
-    :param autoconfig: use snakebite's automatically configured client
-    :type autoconfig: bool
+    Security modes can also be configured by defining appropriate value for
+    the `hadoop.security.authentication` key in `pars`. Kerberos is used
+    automatically if Airflow has been configured to use kerberos.
+
+    :param str hdfs_conn_id: Connection ID to fetch parameters from.
+    :param bool autoconf: Whether to use autoconfig to discover
+        configuration options from the hdfs XML configuration files.
     """
-    def __init__(self, hdfs_conn_id='hdfs_default', proxy_user=None,
-                 autoconfig=False):
-        if not snakebite_imported:
-            raise ImportError(
-                'This HDFSHook implementation requires snakebite, but '
-                'snakebite is not compatible with Python 3 '
-                '(as of August 2015). Please use Python 2 if you require '
-                'this hook  -- or help by submitting a PR!')
+
+    def __init__(self, hdfs_conn_id=None):
```

Comment:
_I added a few deprecation utility classes/functions to help keep the re-written classes as backwards compatible as possible._

This is not the case here: `__init__(self, hdfs_conn_id='hdfs_default', proxy_user=None, autoconfig=False) -> __init__(self, hdfs_conn_id=None)`